### ISSUE TITLE
elm-watch hangs on error in dependency stage

### ISSUE THREAD

**Issue by BrianHicks** [lydell/elm-watch#94]
Hey Simon! I have a weird error here: it looks like `elm-watch` hangs when `elm make` exits 1. Similar to #92, I can't seem to reproduce it outside of my project—it just won't work in a simple SSCCE. However, I was able to stick enough `console.log` statements in the code that gets run to know that it's hanging somewhere in/after [the main `then` handler that sets up the ctrl-c hook](https://github.com/lydell/elm-watch/blob/f6f6bd670b52aa21852f4a12308c434ca126fc60/src/index.ts#L120-L140). Seems like there are tasks still running indefinitely in my case. I was able to use `why-is-node-running` to get a list of tasks that are still running, which I can share privately with you if it'd be helpful.

Anyway, one thing that reliably resolves the issue is to modify that code to look something like this:

```diff
   elmWatchCli(process.argv.slice(2)).then((exitCode) => {
     process.exitCode = exitCode;
     if (process.stdout.isTTY) {
       process.stdout.write(
         "Exiting elm-watch. Press ctrl+c (again) to force."
       );
       process.once("exit", () => {
         process.stdout.cursorTo(0);
         process.stdout.clearLine(0);
       });
-    }
+    } else {
+      process.exit(exitCode)
+    }
   }).catch((error) => {
```

Would you be open to such a change or is there some important thing in the event loop that needs to be cleaned up before the process can exit?

Versions: node v20.9.0 and elm-watch 2.0.0-beta.2

**Comment by lydell** [comment]
Hi! The problem with `process.exit` is that it can exit the process before Node.js has flushed all text that is being printed to stdout, so you can end up with truncated output.

One thing I’ve found useful when debugging things like this, is to write a small Node.js script (or bash script or whatever) called `elm` that pretends to be `elm` but does fake things to force certain scenarios. That’ll test if the logic in elm-watch is correct. If not that’s easy to fix. But maybe more likely it’s something weird Node.js specific that is going on.

Could you write a bit about how this bug impacts you?

**Comment by BrianHicks** [comment]
> Could you write a bit about how this bug impacts you?

It results in `elm-watch make` hanging in CI for like 30 minutes until GitHub Actions reaches a timeout and kills the job. Not too bad in individual PRs, but it holds up our merge queue and causes a lot of rebuilds. It's reached the point where it's a notable pain point for folks working in the repo.

We had a hard time isolating this to `elm-watch`, but it happens with or without our build/test harness in this repo.

> One thing I’ve found useful when debugging things like this, is to write a small Node.js script (or bash script or whatever) called elm that pretends to be elm but does fake things to force certain scenarios. That’ll test if the logic in elm-watch is correct. If not that’s easy to fix.

I forgot to say, but I did this. Here's my script including the particular error that consistently hangs:

```bash
#!/usr/bin/env bash

cat > /dev/stderr <<EOF
-- PROBLEM DOWNLOADING PACKAGE ------------------------------------------------

I was trying to download the source code for wolfadex/elm-calendar 1.0.1, so I
tried to fetch:

    https://github.com/wolfadex/elm-calendar/zipball/1.0.1/

But it came back as 500 Internal Server Error

This may mean some online endpoint changed in an unexpected way, so if does not
seem like something on your side is causing this (e.g. firewall) please report
this to https://github.com/elm/compiler/issues with your operating system, Elm
version, the command you ran, the terminal output, and any additional
information that can help others reproduce the error!
EOF

exit 1
```

> But maybe more likely it’s something weird Node.js specific that is going on.

It's definitely something node.js-specific. In a SSCCE project it does not happen. There are a few layers of indirection in our node setup (we use Volta to manage node versions) so it's likely somewhere in there, but… it's happening.

I cleaned up private paths from the output of `why-is-node-running`. Here's the output:

<details>

```
There are 6 handle(s) keeping the process running

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:53
node:internal/bootstrap/switches/is_main_thread:172
MY_PROJECT/node_modules/.pnpm/siginfo@2.0.0/node_modules/siginfo/index.js:3                           - if (process.stderr.isTTY || force === true) {
MY_PROJECT/node_modules/.pnpm/why-is-node-running@2.2.2/node_modules/why-is-node-running/include.js:3 - require('siginfo')(why, true)
node:internal/modules/cjs/loader:1241
node:internal/modules/cjs/loader:1295
node:internal/modules/cjs/loader:1091

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:198
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9358 - stdin = process.stdin,
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9437 - elmWatchCli(process.argv.slice(2)).then((exitCode) => {
node:internal/modules/cjs/loader:1241
node:internal/modules/cjs/loader:1295
node:internal/modules/cjs/loader:1091
node:internal/modules/cjs/loader:938

# TTYWRAP
node:internal/async_hooks:202
node:internal/bootstrap/switches/is_main_thread:53
node:internal/bootstrap/switches/is_main_thread:150
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9359 - stdout = process.stdout,
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9437 - elmWatchCli(process.argv.slice(2)).then((exitCode) => {
node:internal/modules/cjs/loader:1241
node:internal/modules/cjs/loader:1295
node:internal/modules/cjs/loader:1091

# WORKER
node:internal/async_hooks:202
node:internal/worker:214
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:3357 - this.worker = new import_worker_threads.Worker(path5.join(__dirname, "PostprocessWorker.js"), {
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:3321 - const newWorker = new PostprocessWorker(
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9273 - postprocessWorkerPool.getOrCreateAvailableWorker();
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9404 - result = await run3(
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9423 - doIt().then(resolve3).catch(reject).finally(() => {
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9400 - return new Promise((resolve3, reject) => {

# MESSAGEPORT
node:internal/async_hooks:202
node:internal/worker:214
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:3357 - this.worker = new import_worker_threads.Worker(path5.join(__dirname, "PostprocessWorker.js"), {
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:3321 - const newWorker = new PostprocessWorker(
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9273 - postprocessWorkerPool.getOrCreateAvailableWorker();
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9404 - result = await run3(
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9423 - doIt().then(resolve3).catch(reject).finally(() => {
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9400 - return new Promise((resolve3, reject) => {

# MESSAGEPORT
node:internal/async_hooks:202
node:internal/worker:254
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:3357 - this.worker = new import_worker_threads.Worker(path5.join(__dirname, "PostprocessWorker.js"), {
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:3321 - const newWorker = new PostprocessWorker(
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9273 - postprocessWorkerPool.getOrCreateAvailableWorker();
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9404 - result = await run3(
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9423 - doIt().then(resolve3).catch(reject).finally(() => {
MY_PROJECT/node_modules/.pnpm/elm-watch@2.0.0-beta.2/node_modules/elm-watch/index.js:9400 - return new Promise((resolve3, reject) => {
```

</details>

I'm not sure how to interet the things that are still running, but to me it looks like:

- stdin is still open
- stdout is still open
- the postprocessing worker is still alive

Not sure why those are cleaned up on a different version of node, though!

> The problem with process.exit is that it can exit the process before Node.js has flushed all text that is being printed to stdout, so you can end up with truncated output.

Ah! Would it be possible to flush that explicitly, then?

**Comment by lydell** [comment]
I managed to reproduce now in the `example/` directory with your bash script as a fake `elm`. I should be able to track it down now. Thanks!

**Comment by lydell** [comment]
Fix released in v1.1.3 and v2.0.0-beta.3.

**Comment by BrianHicks** [comment]
Amazing, thank you so much!

Looks like postprocessing was indeed the difference between my main repo and the SSCCE that couldn’t reproduce the example.

We’ve got this in production now and we’ll keep an eye out to see if there are any other hangs!
